// tream/src/stream.ts
export interface Pull<Type> {
    (push: Push<Type>, done?: Done): Undo<Type>;
}

export interface Push<Type> {
    (val: Type, pull: Pull<Type>): void;
}

export interface Done {
    (): void;
}

export interface Undo<Type> {
    (): Pull<Type>;
}

export interface Send<Type> {
    (val: Type): void;
}

export type Sink<Type> = [Send<Type>, Done];

export interface Fork<Type> {
    (): Pull<Type>;
}

export class None { };

export const none: None = new None();
export type Maybe<Type> = Type | None;

export function some<Type>(v: Maybe<Type>): v is Type {
    return v !== none;
}

const defer: (fn: () => void) => (() => void) =
    typeof setImmediate == 'function' ? fn => {
        const ctx = setImmediate(fn);
        return () => { clearImmediate(ctx); };
    } : fn => {
        const ctx = setTimeout(fn, 0);
        return () => { clearTimeout(ctx); };
    };

export function id<Type>(val: Type): () => Type {
    return () => val;
}

export function once<Type>(val: Type): Pull<Type> {
    return iterate([val]);
}

export function empty<Type>(push: Push<Type>, done: Done): Undo<Type> {
    const cancel = defer(done);
    return () => {
        cancel();
        return empty;
    };
}

export function never<Type>(push: Push<Type>, done: Done): Undo<Type> {
    return () => {
        return never;
    };
}

export function repeat<Type>(val: Type): Pull<Type> {
    const pull_ = (push: Push<Type>, done: Done) => {
        const cancel = defer(() => {
            push(val, pull_);
        });
        return () => {
            cancel();
            return pull_;
        };
    };
    return pull_;
}

export function map<Type, RType>(fn: (val: Type) => RType, src: Pull<Type>): Pull<RType> {
    return filter_map(fn, src);
}

export function filter<Type>(fn: (val: Type) => boolean, src: Pull<Type>): Pull<Type> {
    return filter_map<Type, Type>(val => fn(val) ? val : none, src);
}

export function filter_map<Type, RType>(fn: (val: Type) => Maybe<RType>, src: Pull<Type>): Pull<RType> {
    return (push: Push<RType>, done: Done) => {
        const push_ = (val: Type, src: Pull<Type>) => {
            const rval = fn(val);
            if (rval !== none) {
                push(rval as RType, filter_map(fn, src));
            } else {
                undo = src(push_, done);
            }
        };
        let undo = src(push_, done);
        return () => filter_map(fn, undo());
    };
}

export function scan<State, Type, RType>(s: State, fn: (s: State, val: Type) => [State, Maybe<RType>], src: Pull<Type>): Pull<RType> {
    return filter_map((val: Type) => {
        const [new_s, rval] = fn(s, val);
        s = new_s;
        return rval;
    }, src);
}

export function forward<Type>(fn: () => Pull<Type>, src: Pull<Type>): Pull<Type> {
    let pull_ = (push: Push<Type>, done: Done) => {
        let undo = src((val: Type, src_: Pull<Type>) => {
            src = src_;
            push(val, pull_);
        }, () => {
            src = fn();
            pull_ = (push: Push<Type>, done: Done) => {
                const undo = src((val: Type, src_: Pull<Type>) => {
                    push(val, src_);
                }, done);
                return () => {
                    src = undo();
                    return pull_;
                };
            };
            undo = pull_(push, done);
        });
        return () => {
            src = undo();
            return pull_;
        };
    };
    return pull_;
}

export function then<Type, RType>(fn: (val: Type) => Pull<RType>, src: Pull<Type>): Pull<RType> {
    let _src: Maybe<Pull<RType>> = none;
    const pull_ = (push: Push<RType>, done: Done) => {
        let undo_src: Maybe<Undo<Type>> = none;
        let undo_res: Maybe<Undo<RType>> = none;
        const get_res = () => {
            undo_res = (_src as Pull<RType>)((val: RType, src_: Pull<RType>) => {
                undo_res = none;
                _src = src_;
                push(val, pull_);
            }, () => {
                undo_res = none;
                _src = none;
                get_src();
            });
        };
        const get_src = () => {
            undo_src = src((val: Type, src_: Pull<Type>) => {
                undo_src = none;
                src = src_;
                _src = fn(val);
                get_res();
            }, done);
        };
        if (_src !== none) {
            get_res();
        } else {
            get_src();
        }
        return () => {
            if (undo_src !== none) {
                src = (undo_src as Undo<Type>)();
            }
            if (undo_res !== none) {
                _src = (undo_res as Undo<RType>)();
            }
            return pull_;
        };
    };
    return pull_;
}

export function each<Type, RType>(fn: (val: Type) => Pull<RType>, src: Pull<Type>): Pull<RType> {
    let _src: Maybe<Pull<RType>> = none;
    const pull_ = (push: Push<RType>, done: Done) => {
        let undo_src: Maybe<Undo<Type>> = none;
        let undo_res: Maybe<Undo<RType>> = none;
        const get_res = () => {
            undo_res = (_src as Pull<RType>)((val: RType, src_: Pull<RType>) => {
                _src = src_;
                src = (undo_src as Undo<Type>)();
                push(val, pull_);
            }, () => {
                undo_res = none;
                _src = none;
            });
        };
        const get_src = () => {
            undo_src = src((val: Type, src_: Pull<Type>) => {
                src = src_;
                if (undo_res !== none) {
                    (undo_res as Undo<RType>)();
                }
                _src = fn(val);
                get_src();
                get_res();
            }, () => {
                if (undo_res !== none) {
                    (undo_res as Undo<RType>)();
                }
                done();
            });
        };
        get_src();
        if (_src !== none) {
            get_res();
        }
        return () => {
            src = (undo_src as Undo<Type>)();
            if (undo_res !== none) {
                _src = (undo_res as Undo<RType>)();
            }
            return pull_;
        };
    };
    return pull_;
}

export function take<Type>(fn: (val: Type) => boolean, src: Pull<Type>): Pull<Type> {
    return (push: Push<Type>, done: Done) => {
        const undo = src((val: Type, src: Pull<Type>) => {
            if (fn(val)) {
                push(val, take(fn, src));
            } else {
                done();
            }
        }, done);
        return () => take(fn, undo());
    };
}

export function head<State, Type>(st: State, fn: (s: State) => [State, boolean], src: Pull<Type>): Pull<Type> {
    const [st_, res] = fn(st);
    return res ? (push: Push<Type>, done: Done): Undo<Type> => {
        const undo = src((val: Type, src: Pull<Type>) => {
            push(val, head(st_, fn, src));
        }, done);
        return () => head(st, fn, undo());
    } : empty;
}

export function taken<Type>(n: number, src: Pull<Type>): Pull<Type> {
    return head(n, n => n > 0 ? [n - 1, true] : [n, false], src);
}

export function skip<Type>(fn: (val: Type) => boolean, src: Pull<Type>): Pull<Type> {
    return (push: Push<Type>, done: Done) => {
        const push_ = (val: Type, src: Pull<Type>) => {
            if (fn(val)) {
                undo = src(push_, done);
            } else {
                push(val, src);
            }
        };
        let undo = src(push_, done);
        return () => skip(fn, undo());
    };
}

export function skipn<Type>(n: number, src: Pull<Type>): Pull<Type> {
    return skip(val => n > 0 ? (--n, true) : false, src);
}

export function fold<Type, RType>(out: RType, fn: (out: RType, val: Type) => RType, src: Pull<Type>): Pull<RType> {
    const pull_ = (push: Push<RType>, done: Done) => {
        const done_ = () => {
            push(out, empty);
        };
        const push_ = (val: Type, src: Pull<Type>) => {
            out = fn(out, val);
            undo = src(push_, done_);
        };
        let undo = src(push_, done_);
        return () => (src = undo(), pull_);
    };
    return pull_;
}

export function collect<Type>(src: Pull<Type>): Pull<Type[]> {
    return fold([], (out: Type[], val: Type) => (out.push(val), out), src);
}

export function last<Type>(src: Pull<Type>): Pull<Maybe<Type>> {
    return fold(none, (out: Maybe<Type>, val: Type) => val, src);
}

export function select<Type>(srcs: Pull<Type>[]): Pull<Type> {
    if (srcs.length > 0) {
        const pull_ = (push: Push<Type>, done: Done) => {
            const dones: number[] = [];
            const undos = srcs.map((src, src_idx) => src((val: Type, src: Pull<Type>) => {
                srcs = undos
                    .map((undo, idx) => idx != src_idx ?
                        (dones.indexOf(idx) == -1 ? undo() : empty) : src)
                    .filter(src => src != empty);
                push(val, srcs.length > 0 ? pull_ : empty);
            }, () => {
                dones.push(src_idx);
                if (dones.length == undos.length) {
                    done();
                }
            }));
            return () => {
                srcs = undos.map(undo => undo());
                return pull_;
            };
        };

        return pull_;
    } else {
        return empty;
    }
}

function nones<Type>(n: number): Maybe<Type>[] {
    const a = new Array(n);
    for (let i = 0; i < n; i++) {
        a[i] = none;
    }
    return a;
}

export function combine<Type>(srcs: Pull<Type>[]): Pull<Maybe<Type>[]>;
export function combine<Type1>(srcs: [Pull<Type1>]): Pull<[Maybe<Type1>]>;
export function combine<Type1, Type2>(srcs: [Pull<Type1>, Pull<Type2>]): Pull<[Maybe<Type1>, Maybe<Type2>]>;
export function combine<Type1, Type2, Type3>(srcs: [Pull<Type1>, Pull<Type2>, Pull<Type3>]): Pull<[Maybe<Type1>, Maybe<Type2>, Maybe<Type3>]>;
export function combine<Type1, Type2, Type3, Type4>(srcs: [Pull<Type1>, Pull<Type2>, Pull<Type3>, Pull<Type4>]): Pull<[Maybe<Type1>, Maybe<Type2>, Maybe<Type3>, Maybe<Type4>]>;
export function combine<Type1, Type2, Type3, Type4, Type5>(srcs: [Pull<Type1>, Pull<Type2>, Pull<Type3>, Pull<Type4>, Pull<Type5>]): Pull<[Maybe<Type1>, Maybe<Type2>, Maybe<Type3>, Maybe<Type4>, Maybe<Type5>]>;
export function combine<Type1, Type2, Type3, Type4, Type5, Type6>(srcs: [Pull<Type1>, Pull<Type2>, Pull<Type3>, Pull<Type4>, Pull<Type5>, Pull<Type6>]): Pull<[Maybe<Type1>, Maybe<Type2>, Maybe<Type3>, Maybe<Type4>, Maybe<Type5>, Maybe<Type6>]>;
export function combine<Type1, Type2, Type3, Type4, Type5, Type6, Type7>(srcs: [Pull<Type1>, Pull<Type2>, Pull<Type3>, Pull<Type4>, Pull<Type5>, Pull<Type6>, Pull<Type7>]): Pull<[Maybe<Type1>, Maybe<Type2>, Maybe<Type3>, Maybe<Type4>, Maybe<Type5>, Maybe<Type6>, Maybe<Type7>]>;
export function combine<Type1, Type2, Type3, Type4, Type5, Type6, Type7, Type8>(srcs: [Pull<Type1>, Pull<Type2>, Pull<Type3>, Pull<Type4>, Pull<Type5>, Pull<Type6>, Pull<Type7>, Pull<Type8>]): Pull<[Maybe<Type1>, Maybe<Type2>, Maybe<Type3>, Maybe<Type4>, Maybe<Type5>, Maybe<Type6>, Maybe<Type7>, Maybe<Type8>]>;
export function combine<Type1, Type2, Type3, Type4, Type5, Type6, Type7, Type8, Type9>(srcs: [Pull<Type1>, Pull<Type2>, Pull<Type3>, Pull<Type4>, Pull<Type5>, Pull<Type6>, Pull<Type7>, Pull<Type8>, Pull<Type9>]): Pull<[Maybe<Type1>, Maybe<Type2>, Maybe<Type3>, Maybe<Type4>, Maybe<Type5>, Maybe<Type6>, Maybe<Type7>, Maybe<Type8>, Maybe<Type9>]>;
export function combine<Type1, Type2, Type3, Type4, Type5, Type6, Type7, Type8, Type9, Type10>(srcs: [Pull<Type1>, Pull<Type2>, Pull<Type3>, Pull<Type4>, Pull<Type5>, Pull<Type6>, Pull<Type7>, Pull<Type8>, Pull<Type9>, Pull<Type10>]): Pull<[Maybe<Type1>, Maybe<Type2>, Maybe<Type3>, Maybe<Type4>, Maybe<Type5>, Maybe<Type6>, Maybe<Type7>, Maybe<Type8>, Maybe<Type9>, Maybe<Type10>]>;
export function combine<Type1, Type2, Type3, Type4, Type5, Type6, Type7, Type8, Type9, Type10, Type11>(srcs: [Pull<Type1>, Pull<Type2>, Pull<Type3>, Pull<Type4>, Pull<Type5>, Pull<Type6>, Pull<Type7>, Pull<Type8>, Pull<Type9>, Pull<Type10>, Pull<Type11>]): Pull<[Maybe<Type1>, Maybe<Type2>, Maybe<Type3>, Maybe<Type4>, Maybe<Type5>, Maybe<Type6>, Maybe<Type7>, Maybe<Type8>, Maybe<Type9>, Maybe<Type10>, Maybe<Type11>]>;
export function combine<Type1, Type2, Type3, Type4, Type5, Type6, Type7, Type8, Type9, Type10, Type11, Type12>(srcs: [Pull<Type1>, Pull<Type2>, Pull<Type3>, Pull<Type4>, Pull<Type5>, Pull<Type6>, Pull<Type7>, Pull<Type8>, Pull<Type9>, Pull<Type10>, Pull<Type11>, Pull<Type12>]): Pull<[Maybe<Type1>, Maybe<Type2>, Maybe<Type3>, Maybe<Type4>, Maybe<Type5>, Maybe<Type6>, Maybe<Type7>, Maybe<Type8>, Maybe<Type9>, Maybe<Type10>, Maybe<Type11>, Maybe<Type12>]>;
export function combine<Type>(srcs: Pull<Type>[]): Pull<Maybe<Type>[]> {
    return scan(nones<Type>(srcs.length), (vals, [idx, val]) => {
        vals = vals.slice(0);
        vals[idx] = val;
        return [vals, vals];
    }, select(srcs.map((src, idx) => map(val => <[number, Type]>[idx, val], src))));
}

function has_nones<Type>(vals: Maybe<Type>[]): boolean {
    for (const val of vals) {
        if (val === none) {
            return true;
        }
    }
    return false;
}

export function join<Type>(srcs: Pull<Type>[]): Pull<Type[]>;
export function join<Type1, Type2>(srcs: [Pull<Type1>, Pull<Type2>]): Pull<[Type1, Type2]>;
export function join<Type1, Type2, Type3>(srcs: [Pull<Type1>, Pull<Type2>, Pull<Type3>]): Pull<[Type1, Type2, Type3]>;
export function join<Type1, Type2, Type3, Type4>(srcs: [Pull<Type1>, Pull<Type2>, Pull<Type3>, Pull<Type4>]): Pull<[Type1, Type2, Type3, Type4]>;
export function join<Type1, Type2, Type3, Type4, Type5>(srcs: [Pull<Type1>, Pull<Type2>, Pull<Type3>, Pull<Type4>, Pull<Type5>]): Pull<[Type1, Type2, Type3, Type4, Type5]>;
export function join<Type1, Type2, Type3, Type4, Type5, Type6>(srcs: [Pull<Type1>, Pull<Type2>, Pull<Type3>, Pull<Type4>, Pull<Type5>, Pull<Type6>]): Pull<[Type1, Type2, Type3, Type4, Type5, Type6]>;
export function join<Type1, Type2, Type3, Type4, Type5, Type6, Type7>(srcs: [Pull<Type1>, Pull<Type2>, Pull<Type3>, Pull<Type4>, Pull<Type5>, Pull<Type6>, Pull<Type7>]): Pull<[Type1, Type2, Type3, Type4, Type5, Type6, Type7]>;
export function join<Type1, Type2, Type3, Type4, Type5, Type6, Type7, Type8>(srcs: [Pull<Type1>, Pull<Type2>, Pull<Type3>, Pull<Type4>, Pull<Type5>, Pull<Type6>, Pull<Type7>, Pull<Type8>]): Pull<[Type1, Type2, Type3, Type4, Type5, Type6, Type7, Type8]>;
export function join<Type1, Type2, Type3, Type4, Type5, Type6, Type7, Type8, Type9>(srcs: [Pull<Type1>, Pull<Type2>, Pull<Type3>, Pull<Type4>, Pull<Type5>, Pull<Type6>, Pull<Type7>, Pull<Type8>, Pull<Type9>]): Pull<[Type1, Type2, Type3, Type4, Type5, Type6, Type7, Type8, Type9]>;
export function join<Type1, Type2, Type3, Type4, Type5, Type6, Type7, Type8, Type9, Type10>(srcs: [Pull<Type1>, Pull<Type2>, Pull<Type3>, Pull<Type4>, Pull<Type5>, Pull<Type6>, Pull<Type7>, Pull<Type8>, Pull<Type9>, Pull<Type10>]): Pull<[Type1, Type2, Type3, Type4, Type5, Type6, Type7, Type8, Type9, Type10]>;
export function join<Type1, Type2, Type3, Type4, Type5, Type6, Type7, Type8, Type9, Type10, Type11>(srcs: [Pull<Type1>, Pull<Type2>, Pull<Type3>, Pull<Type4>, Pull<Type5>, Pull<Type6>, Pull<Type7>, Pull<Type8>, Pull<Type9>, Pull<Type10>, Pull<Type11>]): Pull<[Type1, Type2, Type3, Type4, Type5, Type6, Type7, Type8, Type9, Type10, Type11]>;
export function join<Type1, Type2, Type3, Type4, Type5, Type6, Type7, Type8, Type9, Type10, Type11, Type12>(srcs: [Pull<Type1>, Pull<Type2>, Pull<Type3>, Pull<Type4>, Pull<Type5>, Pull<Type6>, Pull<Type7>, Pull<Type8>, Pull<Type9>, Pull<Type10>, Pull<Type11>, Pull<Type12>]): Pull<[Type1, Type2, Type3, Type4, Type5, Type6, Type7, Type8, Type9, Type10, Type11, Type12]>;
export function join<Type>(srcs: Pull<Type>[]): Pull<Type[]> {
    return skip(has_nones, combine(srcs)) as any as Pull<Type[]>;
}

export function chain<Type>(srcs: Pull<Type>[]): Pull<Type> {
    if (srcs.length > 0) {
        let src = srcs[0];
        srcs = srcs.slice(1);
        const pull_ = (push: Push<Type>, done: Done) => {
            let undo: Undo<Type>;
            const do_pull = () => {
                undo = src((val: Type, src_: Pull<Type>) => {
                    src = src_;
                    push(val, pull_);
                }, () => {
                    if (srcs.length > 0) {
                        src = srcs[0];
                        srcs = srcs.slice(1);
                        do_pull();
                    } else {
                        done();
                    }
                });
            };
            do_pull();
            return () => {
                src = undo();
                return pull_;
            };
        };
        return pull_;
    } else {
        return empty;
    }
}

interface ForkBranch<Type> {
    sink: Maybe<Sink<Type>>;
    vals: Type[];
};

function fork_has_sink<Type>(branches: ForkBranch<Type>[]): boolean {
    for (const { sink } of branches) {
        if (sink !== none) {
            return true;
        }
    }
    return false;
}

function fork_do_push<Type>(branches: ForkBranch<Type>[], val: Type) {
    for (let branch of branches) {
        const { sink, vals } = branch;
        if (sink !== none) {
            const [send,] = sink as Sink<Type>;
            branch.sink = none;
            send(val);
        } else {
            vals.push(val);
        }
    }
}

function fork_do_done<Type>(branches: ForkBranch<Type>[]) {
    for (let branch of branches) {
        const { sink } = branch;
        if (sink !== none) {
            const [, done] = sink as Sink<Type>;
            branch.sink = none;
            done();
        }
    }
}

export function fork<Type>(src: Pull<Type>): Fork<Type> {
    const branches: ForkBranch<Type>[] = [];

    let is_done = false;
    let undo: Maybe<Undo<Type>> = none;

    const try_pull = () => {
        if (undo === none && fork_has_sink(branches)) {
            undo = src((val: Type, src_: Pull<Type>) => {
                src = src_;
                undo = none;
                fork_do_push(branches, val);
            }, () => {
                is_done = true;
                undo = none;
                fork_do_done(branches);
            });
        }
    };

    const try_undo = () => {
        if (undo !== none && !fork_has_sink(branches)) {
            src = (undo as Undo<Type>)();
            undo = none;
        }
    };

    return () => {
        let branch: ForkBranch<Type>;

        const pull_ = (push: Push<Type>, done: Done) => {
            const { vals } = branch;

            if (vals.length > 0) {
                const cancel = defer(() => {
                    push(vals.shift() as Type, pull_);
                });

                return () => {
                    cancel();
                    return pull_;
                };
            }

            if (is_done) {
                const cancel = defer(done);
                return () => {
                    cancel();
                    return pull_;
                };
            }

            branch.sink = [(val) => push(val, pull_), done];
            try_pull();

            return () => {
                branch.sink = none;
                try_undo();

                return pull_;
            };
        };

        branches.push(branch = { sink: none, vals: [] });

        return pull_;
    };
}

export function forkn<Type>(n: number, src: Pull<Type>): Pull<Type>[] {
    const src_ = fork(src);
    const out: Pull<Type>[] = [];
    for (let i = 0; i < n; i++) {
        out.push(src_());
    }
    return out;
}

export function fork_filter<Type>(fns: ((val: Type) => boolean)[], src: Pull<Type>): Pull<Type>[] {
    return forkn(fns.length, src).map((pull, idx) => filter(fns[idx], pull));
}

export function channel<Type>(): [Sink<Type>, Pull<Type>] {
    let sink: Maybe<Sink<Type>> = none;
    const vals: Type[] = [];

    const send = (val: Type) => {
        if (sink !== none) {
            const [send_,] = sink as Sink<Type>;
            sink = none;
            send_(val);
        } else {
            vals.push(val);
        }
    };

    let is_done = false;
    const done = () => {
        is_done = true;
        if (sink !== none) {
            const [, done_] = sink as Sink<Type>;
            sink = none;
            done_();
        }
    };

    const pull = (push: Push<Type>, done: Done) => {
        if (vals.length > 0) {
            const cancel = defer(() => {
                push(vals.shift() as Type, pull);
            });
            return () => {
                cancel();
                return pull;
            };
        }
        if (is_done) {
            const cancel = defer(done);
            return () => {
                cancel();
                return pull;
            };
        }
        sink = [(val) => push(val, pull), done];
        return () => {
            sink = none;
            return pull;
        };
    };

    return [[send, done], pull];
}

export function generate<State, Type>(st: State, fn: (st: State) => [State, Maybe<Type>]): Pull<Type> {
    const [st_, val] = fn(st);
    if (val !== none) {
        const out = (push: Push<Type>, done: Done) => {
            const cancel = defer(() => {
                push(val as Type, generate(st_, fn));
            });
            return () => {
                cancel();
                return out;
            };
        };
        return out;
    } else {
        return empty;
    }
}

export function iterate<Type>(vals: Type[]): Pull<Type> {
    return generate<number, Type>(0, idx => [idx + 1, idx < vals.length ? vals[idx] : none]);
}
