// tream/stream.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var None = /** @class */ (function () {
    function None() {
    }
    return None;
}());
exports.None = None;
;
exports.none = new None();
function some(v) {
    return v !== exports.none;
}
exports.some = some;
var defer = typeof setImmediate == 'function' ? function (fn) {
    var ctx = setImmediate(fn);
    return function () { clearImmediate(ctx); };
} : function (fn) {
    var ctx = setTimeout(fn, 0);
    return function () { clearTimeout(ctx); };
};
function id(val) {
    return function () { return val; };
}
exports.id = id;
function once(val) {
    return iterate([val]);
}
exports.once = once;
function empty(push, done) {
    var cancel = defer(done);
    return function () {
        cancel();
        return empty;
    };
}
exports.empty = empty;
function never(push, done) {
    return function () {
        return never;
    };
}
exports.never = never;
function repeat(val) {
    var pull_ = function (push, done) {
        var cancel = defer(function () {
            push(val, pull_);
        });
        return function () {
            cancel();
            return pull_;
        };
    };
    return pull_;
}
exports.repeat = repeat;
function map(fn, src) {
    return filter_map(fn, src);
}
exports.map = map;
function filter(fn, src) {
    return filter_map(function (val) { return fn(val) ? val : exports.none; }, src);
}
exports.filter = filter;
function filter_map(fn, src) {
    return function (push, done) {
        var push_ = function (val, src) {
            var rval = fn(val);
            if (rval !== exports.none) {
                push(rval, filter_map(fn, src));
            }
            else {
                undo = src(push_, done);
            }
        };
        var undo = src(push_, done);
        return function () { return filter_map(fn, undo()); };
    };
}
exports.filter_map = filter_map;
function scan(s, fn, src) {
    return filter_map(function (val) {
        var _a = fn(s, val), new_s = _a[0], rval = _a[1];
        s = new_s;
        return rval;
    }, src);
}
exports.scan = scan;
function forward(fn, src) {
    var pull_ = function (push, done) {
        var undo = src(function (val, src_) {
            src = src_;
            push(val, pull_);
        }, function () {
            src = fn();
            pull_ = function (push, done) {
                var undo = src(function (val, src_) {
                    push(val, src_);
                }, done);
                return function () {
                    src = undo();
                    return pull_;
                };
            };
            undo = pull_(push, done);
        });
        return function () {
            src = undo();
            return pull_;
        };
    };
    return pull_;
}
exports.forward = forward;
function then(fn, src) {
    var _src = exports.none;
    var pull_ = function (push, done) {
        var undo_src = exports.none;
        var undo_res = exports.none;
        var get_res = function () {
            undo_res = _src(function (val, src_) {
                undo_res = exports.none;
                _src = src_;
                push(val, pull_);
            }, function () {
                undo_res = exports.none;
                _src = exports.none;
                get_src();
            });
        };
        var get_src = function () {
            undo_src = src(function (val, src_) {
                undo_src = exports.none;
                src = src_;
                _src = fn(val);
                get_res();
            }, done);
        };
        if (_src !== exports.none) {
            get_res();
        }
        else {
            get_src();
        }
        return function () {
            if (undo_src !== exports.none) {
                src = undo_src();
            }
            if (undo_res !== exports.none) {
                _src = undo_res();
            }
            return pull_;
        };
    };
    return pull_;
}
exports.then = then;
function each(fn, src) {
    var _src = exports.none;
    var pull_ = function (push, done) {
        var undo_src = exports.none;
        var undo_res = exports.none;
        var get_res = function () {
            undo_res = _src(function (val, src_) {
                _src = src_;
                src = undo_src();
                push(val, pull_);
            }, function () {
                undo_res = exports.none;
                _src = exports.none;
            });
        };
        var get_src = function () {
            undo_src = src(function (val, src_) {
                src = src_;
                if (undo_res !== exports.none) {
                    undo_res();
                }
                _src = fn(val);
                get_src();
                get_res();
            }, function () {
                if (undo_res !== exports.none) {
                    undo_res();
                }
                done();
            });
        };
        get_src();
        if (_src !== exports.none) {
            get_res();
        }
        return function () {
            src = undo_src();
            if (undo_res !== exports.none) {
                _src = undo_res();
            }
            return pull_;
        };
    };
    return pull_;
}
exports.each = each;
function take(fn, src) {
    return function (push, done) {
        var undo = src(function (val, src) {
            if (fn(val)) {
                push(val, take(fn, src));
            }
            else {
                done();
            }
        }, done);
        return function () { return take(fn, undo()); };
    };
}
exports.take = take;
function head(st, fn, src) {
    var _a = fn(st), st_ = _a[0], res = _a[1];
    return res ? function (push, done) {
        var undo = src(function (val, src) {
            push(val, head(st_, fn, src));
        }, done);
        return function () { return head(st, fn, undo()); };
    } : empty;
}
exports.head = head;
function taken(n, src) {
    return head(n, function (n) { return n > 0 ? [n - 1, true] : [n, false]; }, src);
}
exports.taken = taken;
function skip(fn, src) {
    return function (push, done) {
        var push_ = function (val, src) {
            if (fn(val)) {
                undo = src(push_, done);
            }
            else {
                push(val, src);
            }
        };
        var undo = src(push_, done);
        return function () { return skip(fn, undo()); };
    };
}
exports.skip = skip;
function skipn(n, src) {
    return skip(function (val) { return n > 0 ? (--n, true) : false; }, src);
}
exports.skipn = skipn;
function fold(out, fn, src) {
    var pull_ = function (push, done) {
        var done_ = function () {
            push(out, empty);
        };
        var push_ = function (val, src) {
            out = fn(out, val);
            undo = src(push_, done_);
        };
        var undo = src(push_, done_);
        return function () { return (src = undo(), pull_); };
    };
    return pull_;
}
exports.fold = fold;
function collect(src) {
    return fold([], function (out, val) { return (out.push(val), out); }, src);
}
exports.collect = collect;
function last(src) {
    return fold(exports.none, function (out, val) { return val; }, src);
}
exports.last = last;
function select(srcs) {
    if (srcs.length > 0) {
        var pull_1 = function (push, done) {
            var dones = [];
            var undos = srcs.map(function (src, src_idx) { return src(function (val, src) {
                srcs = undos
                    .map(function (undo, idx) { return idx != src_idx ?
                    (dones.indexOf(idx) == -1 ? undo() : empty) : src; })
                    .filter(function (src) { return src != empty; });
                push(val, srcs.length > 0 ? pull_1 : empty);
            }, function () {
                dones.push(src_idx);
                if (dones.length == undos.length) {
                    done();
                }
            }); });
            return function () {
                srcs = undos.map(function (undo) { return undo(); });
                return pull_1;
            };
        };
        return pull_1;
    }
    else {
        return empty;
    }
}
exports.select = select;
function nones(n) {
    var a = new Array(n);
    for (var i = 0; i < n; i++) {
        a[i] = exports.none;
    }
    return a;
}
function combine(srcs) {
    return scan(nones(srcs.length), function (vals, _a) {
        var idx = _a[0], val = _a[1];
        vals = vals.slice(0);
        vals[idx] = val;
        return [vals, vals];
    }, select(srcs.map(function (src, idx) { return map(function (val) { return [idx, val]; }, src); })));
}
exports.combine = combine;
function has_nones(vals) {
    for (var _i = 0, vals_1 = vals; _i < vals_1.length; _i++) {
        var val = vals_1[_i];
        if (val === exports.none) {
            return true;
        }
    }
    return false;
}
function join(srcs) {
    return skip(has_nones, combine(srcs));
}
exports.join = join;
function chain(srcs) {
    if (srcs.length > 0) {
        var src_1 = srcs[0];
        srcs = srcs.slice(1);
        var pull_2 = function (push, done) {
            var undo;
            var do_pull = function () {
                undo = src_1(function (val, src_) {
                    src_1 = src_;
                    push(val, pull_2);
                }, function () {
                    if (srcs.length > 0) {
                        src_1 = srcs[0];
                        srcs = srcs.slice(1);
                        do_pull();
                    }
                    else {
                        done();
                    }
                });
            };
            do_pull();
            return function () {
                src_1 = undo();
                return pull_2;
            };
        };
        return pull_2;
    }
    else {
        return empty;
    }
}
exports.chain = chain;
;
function fork_has_sink(branches) {
    for (var _i = 0, branches_1 = branches; _i < branches_1.length; _i++) {
        var sink = branches_1[_i].sink;
        if (sink !== exports.none) {
            return true;
        }
    }
    return false;
}
function fork_do_push(branches, val) {
    for (var _i = 0, branches_2 = branches; _i < branches_2.length; _i++) {
        var branch = branches_2[_i];
        var sink = branch.sink, vals = branch.vals;
        if (sink !== exports.none) {
            var send = sink[0];
            branch.sink = exports.none;
            send(val);
        }
        else {
            vals.push(val);
        }
    }
}
function fork_do_done(branches) {
    for (var _i = 0, branches_3 = branches; _i < branches_3.length; _i++) {
        var branch = branches_3[_i];
        var sink = branch.sink;
        if (sink !== exports.none) {
            var _a = sink, done = _a[1];
            branch.sink = exports.none;
            done();
        }
    }
}
function fork(src) {
    var branches = [];
    var is_done = false;
    var undo = exports.none;
    var try_pull = function () {
        if (undo === exports.none && fork_has_sink(branches)) {
            undo = src(function (val, src_) {
                src = src_;
                undo = exports.none;
                fork_do_push(branches, val);
            }, function () {
                is_done = true;
                undo = exports.none;
                fork_do_done(branches);
            });
        }
    };
    var try_undo = function () {
        if (undo !== exports.none && !fork_has_sink(branches)) {
            src = undo();
            undo = exports.none;
        }
    };
    return function () {
        var branch;
        var pull_ = function (push, done) {
            var vals = branch.vals;
            if (vals.length > 0) {
                var cancel_1 = defer(function () {
                    push(vals.shift(), pull_);
                });
                return function () {
                    cancel_1();
                    return pull_;
                };
            }
            if (is_done) {
                var cancel_2 = defer(done);
                return function () {
                    cancel_2();
                    return pull_;
                };
            }
            branch.sink = [function (val) { return push(val, pull_); }, done];
            try_pull();
            return function () {
                branch.sink = exports.none;
                try_undo();
                return pull_;
            };
        };
        branches.push(branch = { sink: exports.none, vals: [] });
        return pull_;
    };
}
exports.fork = fork;
function forkn(n, src) {
    var src_ = fork(src);
    var out = [];
    for (var i = 0; i < n; i++) {
        out.push(src_());
    }
    return out;
}
exports.forkn = forkn;
function fork_filter(fns, src) {
    return forkn(fns.length, src).map(function (pull, idx) { return filter(fns[idx], pull); });
}
exports.fork_filter = fork_filter;
function channel() {
    var sink = exports.none;
    var vals = [];
    var send = function (val) {
        if (sink !== exports.none) {
            var send_ = sink[0];
            sink = exports.none;
            send_(val);
        }
        else {
            vals.push(val);
        }
    };
    var is_done = false;
    var done = function () {
        is_done = true;
        if (sink !== exports.none) {
            var _a = sink, done_ = _a[1];
            sink = exports.none;
            done_();
        }
    };
    var pull = function (push, done) {
        if (vals.length > 0) {
            var cancel_3 = defer(function () {
                push(vals.shift(), pull);
            });
            return function () {
                cancel_3();
                return pull;
            };
        }
        if (is_done) {
            var cancel_4 = defer(done);
            return function () {
                cancel_4();
                return pull;
            };
        }
        sink = [function (val) { return push(val, pull); }, done];
        return function () {
            sink = exports.none;
            return pull;
        };
    };
    return [[send, done], pull];
}
exports.channel = channel;
function generate(st, fn) {
    var _a = fn(st), st_ = _a[0], val = _a[1];
    if (val !== exports.none) {
        var out_1 = function (push, done) {
            var cancel = defer(function () {
                push(val, generate(st_, fn));
            });
            return function () {
                cancel();
                return out_1;
            };
        };
        return out_1;
    }
    else {
        return empty;
    }
}
exports.generate = generate;
function iterate(vals) {
    return generate(0, function (idx) { return [idx + 1, idx < vals.length ? vals[idx] : exports.none]; });
}
exports.iterate = iterate;
